### ISSUE TITLE
gunicorn should not log to a file inside of the container

### ISSUE THREAD

**Issue by GrahamDumpleton** [CTFd/CTFd#719]
**Environment**:

  - CTFd Version/Commit:

master and 2.0.0 branch

  - Operating System:

MacOS X, but using docker so not relevant

  - Web Browser and Version:

N/A

**What happened?**

Access an error logs are not sent to stdout/stderr of the docker container as is regarded as best practice for using container. This means they cannot be captured by any logging system for the container platform the image is being run in.

**What did you expect to happen?**

The gunicorn access and error logs should be sent to stdout/stderr.

**How to reproduce your issue**

Run the docker image using ``docker run`` and send the application requests. Neither access or error log information is displayed in the terminal when you ran ``docker run``, nor are they available via ``docker logs``, nor can they be seen by using ``docker attach`` if container running in daemon mode.

**Any associated stack traces or error logs**

N/A

**Solution**

Suggested that instead of using:

```
exec gunicorn 'CTFd:create_app()' \
    --bind '0.0.0.0:8000' \
    --workers $WORKERS \
    --worker-class 'gevent' \
    --access-logfile "${LOG_FOLDER:-/opt/CTFd/CTFd/logs}/access.log" \
    --error-logfile "${LOG_FOLDER:-/opt/CTFd/CTFd/logs}/error.log"
```

use:

```
ACCESS_LOG=${ACCESS_LOG:-/proc/1/fd/1}
ERROR_LOG=${ERROR_LOG:-/proc/1/fd/2}

exec gunicorn 'CTFd:create_app()' \
    --bind '0.0.0.0:8000' \
    --workers $WORKERS \
    --worker-class 'gevent' \
    --access-logfile "$ACCESS_LOG" \
    --error-logfile "$ERROR_LOG"
```

That is, send access log direct to stdout file for process ID 1 in the container, and error log directory to stderr for process ID 1 in the container. This allows the container platform to capture logs and send it to any aggregated logging system it implements.

Note that this is a breaking change in that previously you could override log directory using ``LOG_FOLDER``. You can't solve this problem and send to stdout/stderr by overriding that though. So would be better to move to generally accepted convention for logging in containers of using stdout/stderr.

Note also that strictly speaking you could not supply ``--error-logfile`` as that will default to sending to stderr anyway, but using that ``/proc/1/fd/2`` trick for it makes it easier to allow overriding as that would be fallback and achieve the same result.

For the purposes of debugging, to preserve old way things worked, you could in the ``docker-compose.yml`` file add environment variables:

```
ACCESS_LOG=/var/log/CTFd/access.log
ERROR_LOG=/var/log/CTFd/error.log
```

but add note to README saying that in production shouldn't rely on that and instead rely on platform to capture logs from stdout/stderr.

**Comment by GrahamDumpleton** [comment]
Better default is probably:

```
# Log to stdout/stderr by default
if [ -n "$LOG_FOLDER" ]; then
    ACCESS_LOG=${LOG_FOLDER}/access.log
    ERROR_LOG=${LOG_FOLDER}/error.log
else
    ACCESS_LOG=/proc/1/fd/1
    ERROR_LOG=/proc/1/fd/2
fi
```

Existing ``LOG_FOLDER`` setting in ``docker-compose.yml`` will result in same behaviour for it. Otherwise will log to stdout/stderr and container platform will capture it.

**Comment by GrahamDumpleton** [comment]
Use of ``gunicorn`` under Kubernetes is a problem when using ``/proc/1/fd/2`` as gunicorn for some unknown reason wants to perform seeks on the file, but you can't do that for raw stderr. Works okay under plain docker, just Kubernetes which is the issue.

```
Error: Error: '/proc/1/fd/2' isn't writable [OSError(29, 'Illegal seek')]
```

Solution is to instead use:

```
# Log to stdout/stderr by default
if [ -n "$LOG_FOLDER" ]; then
    ACCESS_LOG=${LOG_FOLDER}/access.log
    ERROR_LOG=${LOG_FOLDER}/error.log
else
    ACCESS_LOG=-
    ERROR_LOG=-
fi
```

The ``-`` is treated by ``gunicorn`` to mean log to terminal (stdout) and it doesn't try and do a seek in that case.

**Comment by GrahamDumpleton** [comment]
Using ``LOG_FOLDER`` for gunicorn logging is still problematic and should probably just be fixed to log to the terminal so captured from container output, or have separate configuration setting. This is because setting ``LOG_FOLDER`` for purposes of specifying location for ``keys.log``, ``logins.log`` and ``registers.log``, redirects gunicorn output when you really don't want that. Capturing those into a persistent volume by setting ``LOG_FOLDER`` should be completely independent of server monitoring logs.

**Comment by ColdHeat** [comment]
Closed with 92e7be224bf9340ee671bd7ae9933feadab1b4d9

**Comment by GrahamDumpleton** [comment]
The reason I didn't touch the location of the other logs files is that I didn't know if you regarded those as being extra important as far as auditing what happened with the application and actions taken by users. Thus, didn't know if was important to be able to put them in a persistent volume, and not disappear in amongst the error and access log. So what originally suggested may not have been the best way, but didn't attempt to suggest anything else since had moved on and wasn't using the bundle scripts for docker anymore.
